L1Trigger/L1TCaloLayer1/src/UCTLayer1.cc

```diff
@@ -0,0 +1,117 @@
+#include <iostream>
+#include <iomanip>
+#include <stdlib.h>
+#include <stdint.h>
+
+#include "UCTLayer1.hh"
+
+#include "UCTCrate.hh"
+#include "UCTCard.hh"
+#include "UCTRegion.hh"
+#include "UCTTower.hh"
+
+#include "UCTGeometry.hh"
+
+UCTLayer1::UCTLayer1() : uctSummary(0) {
+  UCTGeometry g;
+  for(uint32_t crate = 0; crate < g.getNCrates(); crate++) {
+    crates.push_back(new UCTCrate(crate));
+  }
+}
+
+UCTLayer1::~UCTLayer1() {
+  for(uint32_t i = 0; i < crates.size(); i++) {
+    if(crates[i] != 0) delete crates[i];
+  }
+}
+
+bool UCTLayer1::clearEvent() {
+  for(uint32_t i = 0; i < crates.size(); i++) {
+    if(crates[i] != 0) crates[i]->clearEvent();
+  }
+  return true;
+}
+
+const UCTRegion* UCTLayer1::getRegion(int regionEtaIndex, uint32_t regionPhiIndex) const {
+  if(regionEtaIndex == 0 || regionEtaIndex < -7 || regionEtaIndex > 7 ||
+     regionPhiIndex <= 0 || regionPhiIndex >= 19) {
+    return 0;
+  }
+  // Get (0,0) tower region information
+  UCTGeometry g;
+  UCTRegionIndex r = UCTRegionIndex(regionEtaIndex, regionPhiIndex);
+  UCTTowerIndex t = g.getUCTTowerIndex(r);
+  uint32_t absCaloEta = abs(t.first);
+  uint32_t absCaloPhi = abs(t.second);
+  uint32_t crt = g.getCrate(absCaloEta, absCaloPhi);
+  if(crt >= crates.size()) {
+    std::cerr << "UCTLayer1::getRegion - Crate number is wrong - " << std::hex << crt 
```

Comment:
> I have an issue with this message logger policing. [...] Why does CMS ban std::cerr usage? I understand wanting std::out putting crap in every log file, but rare error reporting can be excused when useful for specific projects in my opinion.

I would say that the reason is exactly this: to actually be able to catch the rare error reports when running in an environment which is a bit more complicated than a user running interactively.
For example, the MessageLogger can be configured to handle messages of different severity and from different sources; and send them to different outputs (cour, cerr, files, etc.).
This was used extensively in the HLT to reduce to a minimum the messages that were actually logged, to avoid flooding the rare error reports with the non-critical warnings and other messages.

An other useful feature which we developed in Run 1 and have yet to resurrect in Run 2 is the possibility of monitoring the "rate" of the various error messages (e.g. in a summary or in a DQM plot), and select a sample of events raising the different errors for offline inspection.